attelo/decoding/mst.py

```diff
@@ -1,88 +1,116 @@
+
 from depparse.graph import Digraph
 from .interface import Decoder
 
 # pylint: disable=too-few-public-methods
 
+# Required to work with depparse, working with a "minimum" score of 1e-100
+# Unless we have more than 1e10 nodes, combined scores shouldn't reach it
+MAX_SCORE = 1e90
+MIN_SCORE = -MAX_SCORE
 
-def get_root(_):
-    """ This is used for the construction of the graph. Since we are
-    using the Chu-Liu Edmonds algorithm in it's dependency parsing
-    incarnation, there should be a node that is the root, i.e.  a node
-    that has no incoming edges. This function is supposed to return
-    that node. For the moment it always returns the first
-    node. Syntactic analysis should be sufficient to provide the real
-    head node.
-    """
-    return '1'
+def _get_root(edus):
+    """ Returns the root node for MST/MSDAG algorithm
 
+        Currently, the EDU in first position
+    """
+    return sorted(edus, key=lambda e:e.span)[0]
 
-def _graph(instances, root='1', use_prob=True):
-    """ instances are quadruplets of the form:
-
-            source, target, probability_of_attachment, relation
+def _graph(instances, use_prob=True):
+    """ Builds a directed graph for instances
 
-        root is the "root" of the graph, that is the node that has no incoming
-        nodes
+        instances are quadruplets of the form:
+            edu_source, edu_target, probability_of_attachment, relation
 
-        returns the Maximum Spanning Tree
+        returns a Digraph
     """
+    
+    root_id = _get_root(set(e for s,t,_,_ in instances for e in (s,t))).id
 
     targets = defaultdict(list)
     labels = dict()
     scores = dict()
 
     for source, target, prob, rel in instances:
-        src = source.id
-        tgt = target.id
-        if tgt == root:
+        src, tgt = source.id, target.id
+
+        # Ignore all edges directed to the root
+        if tgt == root_id:
             continue
-        scores[src, tgt] = prob
+
+        scores[src, tgt] = min(MAX_SCORE, max(MIN_SCORE, logit(prob))) if use_prob else prob
         labels[src, tgt] = rel
-        if use_prob:  # probability scores
-            scores[src, tgt] = log(prob if prob != 0.0 else sys.float_info.min)
         targets[src].append(tgt)
 
     return Digraph(targets,
                    lambda s, t: scores[s, t],
-                   lambda s, t: labels[s, t]).mst()
-
+                   lambda s, t: labels[s, t])
 
-def _list_edges(instances, root='1', use_prob=True):
-    """ instances are quadruplets of the form:
+def _msdag(graph):
+    """ Returns a subgraph of graph (a Digraph) corresponding to its
+        Maximum Spanning Directed Acyclic Graph
 
-            source, target, probability_of_attachment, relation
-
-        root is the "root" of the graph, that is the node that has no incoming
-        nodes
-
-        returns a list of edges for the MST graph
+        Alogrithm is semi-greedy-MSDAG as described in Schluter (2014)
```

Comment:
Try to group together issues that logically belong together.  This is a good practice opportunity.